backend/modules/embedder/tfy_embeddings.py

```diff
@@ -51,51 +53,73 @@ def _requests_retry_session(
+    tfy_api_key: Optional[str] = Field(default=None, alias="api_key")
+    """Automatically inferred from env var `TFY_API_KEY` if not provided."""
+    model_kwargs: Dict[str, Any] = Field(default_factory=dict)
+    model_parameters: Dict[str, Any] = Field(default_factory=dict)
+    tfy_llm_gateway_url: Optional[str] = Field(default=None, alias="llm_gateway_url")
+    """Overwrite for tfy_host for LLM Gateway"""
+    batch_size: Optional[int] = Field(default=EMBEDDER_BATCH_SIZE)
+    """The batch size to use for embedding."""
+    parallel_workers: Optional[int] = Field(default=PARALLEL_WORKERS)
+    """The number of parallel workers to use for embedding."""
+    __private_attributes__ = {"_executor", "_endpoint"}
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+        allow_population_by_field_name = True
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validate that api key and python package exists in environment."""
+        values["tfy_api_key"] = values["tfy_api_key"] or os.getenv("TFY_API_KEY")
+        values["tfy_host"] = values["tfy_host"] or os.getenv("TFY_HOST")
+        values["tfy_llm_gateway_url"] = values["tfy_llm_gateway_url"] or os.getenv(
+            "TFY_LLM_GATEWAY_URL"
+        )
+        if not values["tfy_api_key"]:
+            raise ValueError(
+                f"Did not find `tfy_api_key`, please add an environment variable"
+                f" `TFY_API_KEY` which contains it, or pass"
+                f"  `tfy_api_key` as a named parameter."
             )
-            response.raise_for_status()
-            models = response.json()
-            if len(models) == 0:
-                raise ValueError("No model is deployed in the model server")
-            model_names = [m["name"] for m in models]
-        except Exception as ex:
-            raise Exception(f"Error raised by Inference API: {ex}") from ex
-        else:
-            model_name = model_names[0]
-            self.endpoint = (
-                f"{endpoint_url.strip('/')}/v2/models/{model_name}/infer/simple"
+        if not values["tfy_host"]:
+            raise ValueError(
+                f"Did not find `tfy_host`, please add an environment variable"
+                f" `TFY_HOST` which contains it, or pass"
+                f"  `tfy_host` as a named parameter."
             )
-        self.client = None
-        self.batch_size = int(batch_size)
-        self.parallel_workers = int(parallel_workers)
+        return values
+
+    def _init_private_attributes(self):
         self._executor = concurrent.futures.ThreadPoolExecutor(
             max_workers=self.parallel_workers
         )
+        if self.tfy_llm_gateway_url:
+            self._endpoint = f"{self.tfy_llm_gateway_url}/api/inference/embedding"
+        else:
+            self._endpoint = f"{self.tfy_host}/api/llm/api/inference/embedding"
```

Comment:
minor, use urllib urljoin to avoid issues with extra trailing slash on `tfy_llm_gateway_url` / `tfy_host`

Reply:
Since, `tfy_llm_gateway_url` can be of path based routing type, this would not work?